# Pipeline:
- **Query the odb database**: Find query in database (orthoDB) and retrieve homologous proteins. - `database_v6.py`
    - Can choose different groups at different levels. You can choose the level based on name (e.g. “Eukaryota”), target number of sequences, or the group composed of the greatest number of subspecies.
    - Currently, I am choosing the level based on name, and running pipeline on 5 groups for each gene: ['Eukaryota', 'Metazoa', 'Vertebrata', 'Tetrapoda', 'Mammalia']
    - finding the query gene within the database is done using the class `orthoDB_query` in `database_v6.py`
    - the `orthoDB_query` class is then passed as an argument and modified for the next steps in the pipeline
- **Filtering** - `filters.py`
    - Filter out sequences with “X” or “*” characters
    - Filter out sequences that are shorter than a specified fraction of the query sequence length - `filters.py`
        - Currently 0.5 or 0.75
- **Least divergent ortholog**: - `find_LDOs.py`
    - For each organism, find least divergent ortholog (LDO) from the set of paralogs. The user can choose from different methods to do this:
        - msa: Aligns all of the sequences in the ortholog group with MAFFT (v X.XX). For each organism, it chooses the paralog with the highest percent identity (PID) to the query sequence as the LDO. PID is calculated as...
        - msa-by-organism: performs a separate alignment of the sequences in each organism with the query sequence using MAFFT. For each organism, it chooses the paralog with the highest PID to the query sequence as the LDO.
        - alfpy: Uses an alignment free word-based method to calculate the distance between the query sequence and each paralog. For each organism, it chooses the paralog with the highest similarity to the query sequence as the LDO.
- **Cluster**: - `cluster.py`
    - Cluster the group into clusters of >= 90% identity with CD-HIT
    - a list of representative sequences can be provided such that those sequences are preferentially chosen to represent the cluster that they are in
- **Save files**: - `database_v6.py`
    - Save files to a directory with a main folder for each gene and subfolders for each orthogroup level
    - Main information about each group is saved in a json file within that folder
- **main pipeline** - `main_pipeline.py`
    - This runs the above pipeline for a given gene and input parameters (provided as a dictionary or json file)